File: kakao/level2/수식최대화.kt
class Solution {
       fun solution(expression: String): Long {
        var answer: Long = 0
        var orders = arrayOf("*+-","*-+","+*-","-*+","+-*","-+*")
        var switch = arrayListOf(false,false,false)
        orders.forEach{order->
            var curAnswer:Long = 0
            switch[0] = false
            expression.split(order[0]).forEach{ex->
                var curEx = 0L
                switch[1] = false
                ex.split(order[1]).forEach{ex2->
                    var curEx2 = 0L
                    switch[2] = false
                    ex2.split(order[2]).forEach{ ex3->
                        if(!switch[2]){
                            curEx2 = ex3.toLong()
                            switch[2] = true
                        }
                        else curEx2 = cal(curEx2,ex3.toLong(),order[2])
                    }
                    if(!switch[1]){
                        curEx = curEx2
                        switch[1] = true
                    }
                    else curEx = cal(curEx,curEx2,order[1])
                }
                if(!switch[0]){
                    curAnswer = curEx
                    switch[0] = true
                }
                else curAnswer = cal(curAnswer,curEx,order[0])
            }
            answer = kotlin.math.max(answer, kotlin.math.abs(curAnswer))
        }
        return answer
    }
    fun cal(n1:Long, n2:Long, op:Char):Long{
        return when(op){
            '*' -> {n1*n2}
            '+' -> {n1+n2}
            else -> {n1-n2}
        }
    }
}

File: kakao/level2/수식 최대화.py
from itertools import permutations


def cal(ops, ind, exp):
    if ind == 3:
        return int(exp)
    nums = exp.split(ops[ind])
    result = cal(ops, ind+1, nums[0])
    if len(nums) == 1:
        return result
    for num in nums[1:]:
        if ops[ind] == '*':
            result *= cal(ops,ind+1,num)
        elif ops[ind] == '+':
            result += cal(ops,ind+1,num)
        elif ops[ind] == '-':
            result -= cal(ops,ind+1,num)
    return result


def solution(expression):
    answer = 0
    for order in permutations("*-+",3):
        answer = max(answer, abs(cal(order, 0, expression)))
    return answer

print(solution("100-200*300-500+20"))